presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -76,42 +85,95 @@ public PageProcessor(Optional<PageFilter> filter, List<? extends PageProjection>
 
     public PageProcessorOutput process(ConnectorSession session, DriverYieldSignal yieldSignal, Page page)
     {
-        // limit the scope of the dictionary ids to just one page
-        dictionarySourceIdFunction.reset();
+        AggregatedMemoryContext memoryContext = newSimpleAggregatedMemoryContext();
+        return new PageProcessorOutput(memoryContext::getBytes, process(session, yieldSignal, memoryContext, page));
+    }
+
+    public Iterator<Optional<Page>> process(ConnectorSession session, DriverYieldSignal yieldSignal, AggregatedMemoryContext memoryContext, Page page)
+    {
+        WorkProcessor<Page> processor = process(session, yieldSignal, memoryContext, WorkProcessor.create(new SingletonPage(memoryContext, page)));
+        return processor.yieldingIterator();
+    }
+
+    public WorkProcessor<Page> process(ConnectorSession session, DriverYieldSignal yieldSignal, AggregatedMemoryContext memoryContext, WorkProcessor<Page> pages)
+    {
+        return pages.flatTransform(new PageFilterAndProject(session, yieldSignal, memoryContext));
+    }
+
+    private class SingletonPage
```

Comment:
This class looks very generic, maybe you could extract a generic `SingletonWorkProcessor<T>`?